lib/routes/account.js

```diff
@@ -1021,50 +1048,78 @@ module.exports = function (
-        db.account(Buffer(uid, 'hex'))
-          .then(
-            function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+        if(code.length === 32) {
+          // Hex code length of 32 signifies a tokenId, perform login token verification
+          log.begin('Account.RecoveryLoginVerify', request);
+          db.verifyToken(code, {uid: uid})
+            .then(function () {
+            })
+            .done(
+              function () {
+                reply({})
+              },
+              reply
+            )
+        } else {
+          log.begin('Account.RecoveryEmailVerify', request)
+          db.account(Buffer(uid, 'hex'))
+            .then(
+              function (account) {
+                // If the account is already verified, they may be e.g.
+                // clicking a stale link.  Silently succeed.
+                if (account.emailVerified) {
+                  log.increment('account.already_verified')
+                  return true
+                }
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  throw error.invalidVerificationCode()
+                }
+                log.timing('account.verified', Date.now() - account.createdAt)
+                log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                log.increment('account.verified')
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
-            }
-          )
-          .done(
-            function () {
-              reply({})
-            },
-            reply
-          )
+                // send a push notification to all devices that the account changed
+                push.notifyUpdate(uid)
+
+                return db.sessions(account.uid)
+                  .then(function(sessions) {
+                    // TODO How to know which session is associated with the email verification?
+                    // Assuming last created session?
```

Comment:
> In the proposed solution, if email verification code verifies the email, sessionToken and ketFetchToken, wouldn't it verify it for the tokens used on the creation of the account, browser A. So if a user verifies on browser B, browser A session gets verified and logged in?

This seems wrong to me. An verification code should only verify the session/keyFetchToken associated with it, otherwise we are breaking the sandbox the email verification loop is meant to provide.

Here's the scenario I want to prevent:
A determined and patient attacker signs in as a verified user on browser A and waits. The user receives an email and ignores it because they didn't sign in and they didn't understand what led to that email being sent. At a later point, the user starts up a new Fx profile or buys a new Android or iOS device and signs in to Sync. They receive the email, click the verification link. If we verify all tokens, we've just verified the attacker's session too.